File: modules/spread/app/model/ODFCorrector.java
package model;

import java.io.FileOutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;

import org.odftoolkit.odfdom.type.Color;
import org.odftoolkit.simple.SpreadsheetDocument;
import org.odftoolkit.simple.style.Font;
import org.odftoolkit.simple.style.StyleTypeDefinitions.FontStyle;
import org.odftoolkit.simple.table.Cell;
import org.odftoolkit.simple.table.Table;

import play.Logger;

/**
 *
 * @author Stefan Olbrecht
 *
 */
public class ODFCorrector extends SpreadCorrector<SpreadsheetDocument, Table, Cell, Font, Color> {

  private static final String CORRECT_FORMULA = "Formel richtig.";

  private static final int MAXROW = 80;
  private static final int MAXCOLUMN = 22;

  private static final String COLOR_WHITE = "#FFFFFF";
  private static final String FONT = "Arial";
  private static final double FONT_SIZE = 10.;

  @Override
  public void closeDocument(SpreadsheetDocument document) {
    document.close();
  }

  @Override
  public String compareCellFormulas(Cell masterCell, Cell compareCell) {
    String masterFormula = masterCell.getFormula();
    String compareFormula = compareCell.getFormula();
    if (masterFormula == null)
      // Keine Formel zu vergleichen
      return "";
    else if (masterFormula.equals(compareFormula))
      // Formel richtig
      return CORRECT_FORMULA;
    else if (compareFormula == null)
      // Keine Formel von Student angegeben
      return "Formel notwendig.";
    else {
      String diffOfTwoFormulas = HashSetHelper.getDiffOfTwoFormulas(masterFormula, compareFormula);
      if (diffOfTwoFormulas.isEmpty())
	return CORRECT_FORMULA;
      else
	return "Formel falsch. " + diffOfTwoFormulas;
    }
  }

  @Override
  public String compareCellValues(Cell masterCell, Cell compareCell) {
    String masterValue = masterCell.getStringValue();
    String compareValue = compareCell.getStringValue();
    // FIXME: why substring from 0 to first newline?
    if (compareValue.indexOf('\n') != -1)
      compareValue = compareValue.substring(0, compareValue.indexOf('\n'));
    if (compareValue.isEmpty())
      return "Kein Wert angegeben.";
    else if (masterValue.equals(compareValue))
      return "Wert richtig.";
    else
      return "Wert falsch. Erwartet wurde '" + masterValue + "'.";
  }

  @Override
  public String compareChartsInSheet(Table compareSheet, Table sampleSheet) {
    // FIXME: nicht von ODFToolkit unterstützt...
    return null;
  }

  @Override
  public String compareNumberOfChartsInDocument(SpreadsheetDocument compare, SpreadsheetDocument sample) {
    int sampleCount = sample.getChartCount();
    int compareCount = compare.getChartCount();
    if (sampleCount == 0)
      return "Es sollten keine Diagramme erstellt werden.";
    else if (sampleCount != compareCount)
      return "Falsche Anzahl Diagramme im Dokument (Erwartet: " + sampleCount + ", Gefunden: " + compareCount + ").";
    else
      return "Richtige Anzahl Diagramme gefunden.";
  }

  @Override
  public void compareSheet(Table sampleTable, Table compareTable, boolean correctConditionalFormating) {
    if (correctConditionalFormating) {
      // NOTICE: Does not work in ODF Toolkit
    }
    // Iterate over colored cells
    List<Cell> range = getColoredRange(sampleTable);
    for (Cell cellMaster : range) {
      int rowIndex = cellMaster.getRowIndex();
      int columnIndex = cellMaster.getColumnIndex();
      Cell cellCompare = compareTable.getCellByPosition(columnIndex, rowIndex);

      if (cellCompare == null)
	// TODO: Fehler werfen? Kann das überhaupt passieren?
	return;

      // Compare cell values
      String cellValueResult = compareCellValues(cellMaster, cellCompare);
      String cellFormulaResult = compareCellFormulas(cellMaster, cellCompare);

      setCellComment(cellCompare, cellValueResult + "\n" + cellFormulaResult);

      // FIXME: use enum instead of String!
      if ("Wert richtig.".equals(cellValueResult)
          && (cellFormulaResult.isEmpty() || CORRECT_FORMULA.equals(cellFormulaResult)))
	setCellStyle(cellCompare, new Font(FONT, FontStyle.BOLD, FONT_SIZE), Color.GREEN);
      // cellCompare.setFont(new Font(FONT, FontStyle.BOLD, FONT_SIZE,
      // Color.GREEN));
      else
	setCellStyle(cellCompare, new Font(FONT, FontStyle.ITALIC, FONT_SIZE), Color.RED);
      // cellCompare.setFont(new Font(FONT, FontStyle.ITALIC, FONT_SIZE,
      // Color.RED));
    }
  }

  @Override
  public Cell getCellByPosition(Table table, int column, int row) {
    return table.getCellByPosition(column, row);
  }

  @Override
  @SuppressWarnings("deprecation")
  public List<Cell> getColoredRange(Table master) {
    List<Cell> range = new ArrayList<>();
    for (int row = 0; row < MAXROW; row++) {
      for (int column = 0; column < MAXCOLUMN; column++) {
	Cell oCell = master.getRowByIndex(row).getCellByIndex(column);
	if (!oCell.getCellBackgroundColorString().equals(COLOR_WHITE))
	  range.add(oCell);
      }
    }
    return range;
  }

  @Override
  public Table getSheetByIndex(SpreadsheetDocument document, int sheetIndex) {
    return document.getSheetByIndex(sheetIndex);
  }

  @Override
  public int getSheetCount(SpreadsheetDocument document) {
    return document.getSheetCount();
  }

  @Override
  public SpreadsheetDocument loadDocument(Path path) throws CorrectionException {
    try {
      return SpreadsheetDocument.loadDocument(path.toFile());
    } catch (Exception e) {
      throw new CorrectionException("", "Beim Öffnen eines ODF-Dokuments ist ein Fehler aufgetreten.", e);
    }
  }

  @Override
  public void saveCorrectedSpreadsheet(SpreadsheetDocument compareDocument, Path testPath) {
    // @formatter:off
    String fileNameNew = com.google.common.io.Files.getNameWithoutExtension(testPath.toString()) + CORRECTION_ADD_STRING
        + "." + com.google.common.io.Files.getFileExtension(testPath.toString());
    // @formatter:on
    Path savePath = Paths.get(testPath.getParent().toString(), fileNameNew);
    try {
      if (!savePath.getParent().toFile().exists())
	Files.createDirectories(savePath.getParent());

      FileOutputStream fileOut = new FileOutputStream(savePath.toFile());
      compareDocument.save(fileOut);
      fileOut.close();
    } catch (Exception e) {
      Logger.error(ERROR_MESSAGE_SAVE, e);
    }
  }

  @Override
  public void setCellComment(Cell cell, String message) {
    if (message == null || message.isEmpty())
      return;
    cell.setNoteText(message);
  }

  @Override
  public void setCellStyle(Cell cell, Font font, Color color) {
    font.setColor(color);
    cell.setFont(font);
  }

}

File: modules/spread/app/model/ODFCorrector.scala
package model

import java.io.FileOutputStream
import java.nio.file.{Files, Path, Paths}

import com.google.common.io.{Files => GFiles}
import model.StringConsts._
import org.odftoolkit.odfdom.`type`.Color
import org.odftoolkit.simple.SpreadsheetDocument
import org.odftoolkit.simple.style.Font
import org.odftoolkit.simple.table.{Cell, Table}

import scala.util.Try


object ODFCorrector extends SpreadCorrector[SpreadsheetDocument, Table, Cell, Font, Color] {

  val MAXROW = 80
  val MAXCOLUMN = 22

  val COLOR_WHITE = "#FFFFFF"
  val FONT = "Arial"
  val FONT_SIZE: Double = 10

  override def closeDocument(document: SpreadsheetDocument): Unit = document.close()

  override def compareCellFormulas(masterCell: Cell, compareCell: Cell): (Boolean, String) = Option(masterCell.getFormula) match {
    case None => (true, "")
    case Some(masterFormula) => Option(compareCell.getFormula) match {
      case None => (false, COMMENT_FORMULA_MISSING)
      case Some(compareFormula) =>
        if (masterFormula == compareFormula) (true, COMMENT_FORMULA_CORRECT)
        else {
          val diffOfTwoFormulas = SpreadUtils.getDiffOfTwoFormulas(masterFormula, compareFormula)
          if (diffOfTwoFormulas.isEmpty) (true, COMMENT_FORMULA_CORRECT)
          else (false, s"Formel falsch. $diffOfTwoFormulas")
        }
    }
  }

  override def compareCellValues(masterCell: Cell, compareCell: Cell): (Boolean, String) = {
    //    String masterValue = masterCell.getStringValue()
    //    String compareValue = compareCell.getStringValue()
    //    // FIXME: why substring from 0 to first newline?
    //    if (compareValue.indexOf('\n') != -1)
    //      compareValue = compareValue.substring(0, compareValue.indexOf('\n'))
    //    if (compareValue.isEmpty())
    //      return COMMENT_VALUE_MISSING
    //    else if (masterValue.equals(compareValue))
    //      return COMMENT_VALUE_CORRECT
    //    else
    //          return String.format(COMMENT_VALUE_INCORRECT_VAR, masterValue)
    (false, "TODO...")
  }

  override def compareChartsInSheet(compareSheet: Table, sampleSheet: Table): (Boolean, String) = {
    // FIXME: nicht von ODFToolkit unterstützt...
    null
  }

  override def compareNumberOfChartsInDocument(compare: SpreadsheetDocument, sample: SpreadsheetDocument): (Boolean, String) =
    sample.getChartCount match {
      case 0 => (false, COMMENT_CHART_FALSE)
      case sampleCount =>
        val compareCount = compare.getChartCount
        if (sampleCount == compareCount) (true, COMMENT_CHART_NUM_CORRECT)
        else (false, s"Falsche Anzahl Diagramme im Dokument (Erwartet: $sampleCount, Gefunden: $compareCount).")
    }

  override def compareSheet(sampleTable: Table, compareTable: Table, correctConditionalFormating: Boolean): Unit = {
    //    if (correctConditionalFormating) {
    //      // NOTICE: Does not work in ODF Toolkit
    //    }
    //    // Iterate over colored cells
    //    List<Cell> range = getColoredRange(sampleTable)
    //    for (Cell cellMaster : range) {
    //      int rowIndex = cellMaster.getRowIndex()
    //      int columnIndex = cellMaster.getColumnIndex()
    //      Cell cellCompare = compareTable.getCellByPosition(columnIndex, rowIndex)
    //
    //      if (cellCompare == null)
    //	// TODO: Fehler werfen? Kann das überhaupt passieren?
    //	return
    //
    //      // Compare cell values
    //      String cellValueResult = compareCellValues(cellMaster, cellCompare)
    //      String cellFormulaResult = compareCellFormulas(cellMaster, cellCompare)
    //
    //      setCellComment(cellCompare, cellValueResult + "\n" + cellFormulaResult)
    //
    //      // FIXME: use enum instead of String!
    //      if (COMMENT_VALUE_CORRECT.equals(cellValueResult)
    //          && (cellFormulaResult.isEmpty() || COMMENT_FORMULA_CORRECT.equals(cellFormulaResult)))
    //	setCellStyle(cellCompare, new Font(FONT, FontStyle.BOLD, FONT_SIZE), Color.GREEN)
    //      // cellCompare.setFont(new Font(FONT, FontStyle.BOLD, FONT_SIZE,
    //      // Color.GREEN))
    //      else
    //	setCellStyle(cellCompare, new Font(FONT, FontStyle.ITALIC, FONT_SIZE), Color.RED)
    //      // cellCompare.setFont(new Font(FONT, FontStyle.ITALIC, FONT_SIZE,
    //      // Color.RED))
    //    }
  }

  override def getCellByPosition(table: Table, column: Int, row: Int): Cell = table.getCellByPosition(column, row)

  //  @SuppressWarnings("deprecation")
  override def getColoredRange(master: Table): List[Cell] = {
    //    List<Cell> range = new ArrayList<>()
    //    for (int row = 0 row < MAXROW row++) {
    //      for (int column = 0 column < MAXCOLUMN column++) {
    //	Cell oCell = master.getRowByIndex(row).getCellByIndex(column)
    //	if (!oCell.getCellBackgroundColorString().equals(COLOR_WHITE))
    //	  range.add(oCell)
    //      }
    //    }
    //    return range
    List.empty
  }

  override def getSheetByIndex(document: SpreadsheetDocument, sheetIndex: Int): Table = document.getSheetByIndex(sheetIndex)

  override def getSheetCount(document: SpreadsheetDocument): Int = document.getSheetCount

  override def loadDocument(path: Path): Try[SpreadsheetDocument] = Try(SpreadsheetDocument.loadDocument(path.toFile))

  override def saveCorrectedSpreadsheet(compareDocument: SpreadsheetDocument, testPath: Path): Try[Path] = {
    val tPath = testPath.toString
    val fileNameNew = GFiles.getNameWithoutExtension(tPath) + CORRECTION_ADD_STRING + "." + GFiles.getFileExtension(tPath)
    val savePath = Paths.get(testPath.getParent.toString, fileNameNew)
    Try({
      if (!savePath.getParent.toFile.exists())
        Files.createDirectories(savePath.getParent)

      val fileOut: FileOutputStream = new FileOutputStream(savePath.toFile)
      compareDocument.save(fileOut)
      fileOut.close()

      savePath
    })
  }

  override def setCellComment(cell: Cell, message: String): Unit = {
    if (message != null || message.nonEmpty)
      cell.setNoteText(message)
  }

  override def setCellStyle(cell: Cell, font: Font, color: Color): Unit = {
    font.setColor(color)
    cell.setFont(font)
  }

}
